### PR TITLE
alsa-scarlett-gui: fixed desktop entry

### PR DESCRIPTION
**Pull request by inyourfoss**
Fixed desktop entry of the app so that it shows up in the menu. 

###### Description of changes
In place search and replace was implemented in the postInstall phase using sed.

@SebTM as requested the solution now is all in one file.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (SebTM) [inline pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: From what I remember, can't find the exact comment in one of my MR/or where I participated substituteInPlace is preferred over sed, but will leave it to the merger if it's needed as I don't have merge permissions ✌🏻

**Review by SebTM — APPROVED** [review]
LGTM

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (gnumber13) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: Just to give my 2cents to what speaks for sed over substituteInPlace (as far a I could investigate):

First I think it is not that easily testable. Which you really need for non trivial stuff. sed is a standard unix tool which you can easily test for the correct result on every unix/linux command line.

Secondly you could only use substituteInPlace on the exact match of the entire line (as I could not find any documentation about pattern matching)  which means worst case you would have to change it in the default.nix on every release. sed matches the Line from "Exec=" to the end of the line (not matter what comes after  "Exec=") and replaces that. That should make it more resistant to changes later on.

**Comment by gnumber13** [comment]
Hey @SebTM ,
somehow the check for aarch64-darwin has been running for 3 days.
Will someone have to intervene?
And when do the merges happen usually? 
Greets

**Comment by SebTM** [comment]
@ofborg build alsa-scarlett-gui

**Comment by SebTM** [comment]
> Hey @SebTM , somehow the check for aarch64-darwin has been running for 3 days. Will someone have to intervene? And when do the merges happen usually?

I tried to restart it now.

> 

`but will leave it to the merger if it's needed as I don't have merge permissions ✌🏻`
You can post them also [here](https://discourse.nixos.org/t/prs-already-reviewed/2617/1033) after you got one approval at least (at best from a maintainer), I did that now ✌🏻 

Otherwise you can use an overlay to fix it for for your system before merge: https://nixos.wiki/wiki/Overlays

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: sed also does nothing if it matches nothing unlike substituteInPlace which logs a warning.

Anyway, since we are patching multiple lines it would be easier to follow the change, if it would be a patch

Comment (SebTM) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: btw. I don't think try-exec is really needed as we discarded the %F stuff etc.

**Comment by SuperSandro2000** [comment]
> Hey @SebTM , somehow the check for aarch64-darwin has been running for 3 days. Will someone have to intervene? And when do the merges happen usually? Greets

Can be ignored if it never finishes.

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: Can we remove it then?

Comment (gnumber13) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:28]: I will test if the tryexec key can be removed and report back.

Just to summarize the most preferred way would be to use substituteInPlace to substitute the line with the Exec key exactly and replace that with the desired line?
And how would I go about applying the changes? A new PR?

**Comment by gnumber13** [comment]
FYI: @SuperSandro2000 @SebTM 
I added a new commit that manipulates the desktop entry file using "substituteInPlace".

During testing I realized it would be best to have the absolute paths to the executable in the nix store.
Since in my testing KDE Plasma had issues picking up the desktop entry containing the relative path.

I think that might be the more robust solution.

Cinnamon, Gnome and XFCE were working either way.

**Comment by SebTM** [comment]
For the binary I see a need for certain environments to use full paths about the icon thing I'm unsure I found it a few times but in general they are collected by the standardize output-paths and should work out-of-the-box.

Other than that works for me ✌🏻

**Review by SebTM — APPROVED** [review]

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (SebTM) [inline pkgs/applications/audio/alsa-scarlett-gui/default.nix:30]: This replaces "alsa-scarlett-gui" with "alsa-scarlett-gui"?

Comment (gnumber13) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:30]: Yeah sorry those are remnants from testing that line will be removed. As its not needed.

**Review by SebTM — APPROVED** [review]

File: pkgs/applications/audio/alsa-scarlett-gui/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/audio/alsa-scarlett-gui/default.nix:29]: Intendation is not matching 2 spaces which is written in our contributing guide and also we shouldn't substitute pname wherever possible

Comment (SebTM) [reply pkgs/applications/audio/alsa-scarlett-gui/default.nix:29]: https://github.com/NixOS/nixpkgs/pull/246636

**Comment by SebTM** [comment]
Hey @gnumber13, congratulations on your first contribution/pull-request to NixOS/nixpkgs 😃🎉

**Comment by gnumber13** [comment]
Hey @SebTM Thanks! And thank you for your support :)

**Pull request merged**
